node/subsystem/src/util.rs

```diff
@@ -413,25 +415,67 @@ impl<Spawner: Spawn, Job: JobTrait> Jobs<Spawner, Job> {
 		Ok(())
 	}
 
-	/// Get the next message from any of the underlying jobs.
-	async fn next(&mut self) -> Option<Job::FromJob> {
-		self.outgoing_msgs.next().await.and_then(|(e, _)| match e {
-			StreamYield::Item(e) => Some(e),
-			_ => None,
-		})
+	/// Get the pin projection for the `outgoing_msgs` field
+	///
+	/// From the [pin docs]:
+	///
+	/// > It is actually up to the author of the data structure to decide whether the pinned
+	/// > projection for a particular field turns `Pin<&mut Struct>` into `Pin<&mut Field>`
+	/// > or `&mut Field`. There are some constraints, though, and the most important constraint
+	/// > is _consistency_: every field can be _either_ projected to a pinned reference, _or_
+	/// > have pinning removed as part of the projection. If both are done for the same field,
+	/// > that will likely be unsound!
+	///
+	/// In this case, pinning is structural.
+	///
+	/// ## Considerations
+	///
+	/// 1. The struct must only be `Unpin` if all the structural fields are `Unpin`: ✔
+	/// 2. The destructor of the struct must not move structural fields out of its argument: ✔
+	/// 3. Uphold the `Drop` guarantee: once the struct is pinned, the memory which contains
+	///    the content is not overwritten or deallocated without calling the content's destructors.
+	///    I.e. you may not free or reuse the storage without calling `drop`. ✔
+	/// 4. You must not offer any other operations (i.e. `take`) which could lead to data being
+	///    moved out of the structural fields when your type is pinned: ✔
+	///
+	/// [pin docs]: https://doc.rust-lang.org/std/pin/index.html#projections-and-structural-pinning
+	fn pin_get_outgoing_msgs(self: Pin<&mut Self>) -> Pin<&mut StreamUnordered<mpsc::Receiver<Job::FromJob>>> {
```

Comment:
have you considered using [pin-project](https://github.com/taiki-e/pin-project) or [pin-project-lite](https://github.com/taiki-e/pin-project-lite)?

Reply:
Thanks for linking me to those! No, I'd never heard of either of them.

Reply:
dc9e12c